files/grest/rpc/cached_tables/pool_info_cache_setup.sql

```diff
@@ -0,0 +1,203 @@
+    SET
+        retiring_epoch = _retiring_epoch,
+        unixtime = _unixtime
+    WHERE
+        pool_hash_id = _hash_id
+        AND
+        tx_id < _announced_tx_id;
+END;
+$$;
+
+COMMENT ON FUNCTION grest.pool_info_retire_update IS 'Internal function to update pool_info_cache with new retire status';
+
+
+DROP FUNCTION IF EXISTS grest.pool_info_update CASCADE;
+
+CREATE FUNCTION grest.pool_info_update ()
+    RETURNS TRIGGER
+    LANGUAGE plpgsql
+    AS $$
+DECLARE
+    _latest_pool_info_tx_id bigint;
+    _latest_unixtime_cache bigint;
+    _current_unixtime bigint;
+    _current_pool_update_block_id bigint DEFAULT NULL;
+    rec RECORD;
+BEGIN
+    SELECT COALESCE(MAX(tx_id), 0) INTO _latest_pool_info_tx_id FROM grest.pool_info_cache;
+    SELECT COALESCE(MAX(unixtime), 0) INTO _latest_unixtime_cache FROM grest.pool_info_cache;
+    SELECT EXTRACT(EPOCH FROM NOW()) INTO _current_unixtime;
+    IF (_current_unixtime - _latest_unixtime_cache) > 300 THEN
+        -- Add all new entries in pool_update table older than 5 blocks
+        FOR rec IN (SELECT * FROM public.pool_update AS pu WHERE pu.registered_tx_id > _latest_pool_info_tx_id) LOOP
+            SELECT block_id INTO _current_pool_update_block_id FROM public.tx AS t WHERE t.id = rec.registered_tx_id;
+            IF _current_pool_update_block_id IS NOT NULL AND (NEW.id - _current_pool_update_block_id) > 5 THEN
+                PERFORM grest.pool_info_insert(
+                    rec.id,
+                    rec.registered_tx_id,
+                    rec.hash_id,
+                    rec.active_epoch_no,
+                    rec.vrf_key_hash,
+                    rec.margin,
+                    rec.fixed_cost,
+                    rec.pledge,
+                    rec.reward_addr,
+                    rec.meta_id,
+                    _current_unixtime
+                );
+            END IF;
+        END LOOP;
+        -- Check all new entries in pool_retire table older than 5 blocks
+        FOR rec IN (SELECT * FROM public.pool_retire AS pr WHERE pr.announced_tx_id > _latest_pool_info_tx_id) LOOP
+            SELECT block_id INTO _current_pool_update_block_id FROM public.tx AS t WHERE t.id = rec.announced_tx_id;
+            IF _current_pool_update_block_id IS NOT NULL AND (NEW.id - _current_pool_update_block_id) > 5 THEN
+                PERFORM grest.pool_info_retire_update(
+                    rec.hash_id,
+                    rec.announced_tx_id,
+                    rec.retiring_epoch,
+                    _current_unixtime
+                );
+            END IF;
+        END LOOP;
+    END IF;
+    RETURN NULL;
+END;
+$$;
+
+COMMENT ON FUNCTION grest.pool_info_update IS 'Internal function to insert all new pool updates into pool_info cache table';
+
+
+-- Create pool_info_update trigger
+DROP TRIGGER IF EXISTS pool_info_update_trigger ON public.block;
+
+CREATE TRIGGER pool_info_update_trigger
+    AFTER INSERT ON public.block
+    FOR EACH ROW
```

Comment:
Is it better to use `FOR EACH STATEMENT` here?

From docs (https://www.postgresql.org/docs/9.1/sql-createtrigger.html):

> A trigger that is marked FOR EACH ROW is called once for every row that the operation modifies. For example, a DELETE that affects 10 rows will cause any ON DELETE triggers on the target relation to be called 10 separate times, once for each deleted row. In contrast, a trigger that is marked FOR EACH STATEMENT only executes once for any given operation, regardless of how many rows it modifies (in particular, an operation that modifies zero rows will still result in the execution of any applicable FOR EACH STATEMENT triggers).

Reply:
But can you use the NEW "object" in the called trigger function then?
I use the id of the block that causes the trigger to fire to compare against.

Reply:
I think you can still use it. My understanding is it only affects during-sync execution when you can have batches of blocks inserted, and then `FOR EACH ROW` would execute the procedure for each of the e.g. 5000 blocks in a batch. But after sync, I think we can only get 1 by 1 block inserts so I think there is no difference in that case.